src/Esprima/Utils/JavascriptTextWriter.cs

```diff
@@ -0,0 +1,375 @@
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    protected void WriteSpace()
+    {
+        WriteWhiteSpace(" ");
+    }
+
+    protected void WriteLine()
+    {
+        _writer.WriteLine();
+        WhiteSpaceWrittenSinceLastToken = true;
+    }
+
+    protected void WriteWhiteSpace(string value)
+    {
+        _writer.Write(value);
+        WhiteSpaceWrittenSinceLastToken = true;
+    }
+
+    #endregion
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    protected void ForceRecommendedSpace()
+    {
+        LastTokenFlags |= TokenFlags.TrailingSpaceRecommended;
+    }
+
+    public virtual void WriteEpsilon(TokenFlags flags, in WriteContext context) { }
+
+    #region Identifiers
+
+    protected virtual void StartIdentifier(string value, TokenFlags flags, in WriteContext context)
+    {
+        switch (LastTokenType)
+        {
+            case TokenType.BigIntLiteral:
+            case TokenType.BooleanLiteral:
+            case TokenType.Identifier:
+            case TokenType.Keyword:
+            case TokenType.NullLiteral:
+            case TokenType.NumericLiteral:
+            case TokenType.RegularExpression:
+                WriteSpace();
+                break;
+            case TokenType.EOF:
+            case TokenType.Punctuator:
+            case TokenType.StringLiteral:
+            case TokenType.Template:
+                break;
+            default:
+                throw new InvalidOperationException();
+        }
+    }
+
+    public void WriteIdentifier(string value, TokenFlags flags, in WriteContext context)
+    {
+        StartIdentifier(value, flags, in context);
+        _writer.Write(value);
+        WhiteSpaceWrittenSinceLastToken = false;
+        EndIdentifier(value, flags, in context);
+
+        LastTokenType = TokenType.Identifier;
+        LastTokenFlags = flags;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void WriteIdentifier(string value, in WriteContext context)
+    {
+        WriteIdentifier(value, TokenFlags.None, in context);
+    }
+
+    protected virtual void EndIdentifier(string value, TokenFlags flags, in WriteContext context) { }
+
+    #endregion
+
+    #region Keywords
+
+    protected virtual void StartKeyword(string value, TokenFlags flags, in WriteContext context)
+    {
+        switch (LastTokenType)
+        {
+            case TokenType.BigIntLiteral:
+            case TokenType.BooleanLiteral:
+            case TokenType.Identifier:
+            case TokenType.Keyword:
+            case TokenType.NullLiteral:
+            case TokenType.NumericLiteral:
+            case TokenType.RegularExpression:
+                WriteSpace();
+                break;
+            case TokenType.EOF:
+            case TokenType.Punctuator:
+            case TokenType.StringLiteral:
+            case TokenType.Template:
+                break;
+            default:
+                throw new InvalidOperationException();
```

Comment:
should prefer `EsprimaExceptionHelper` for consistency, argument out of range?

Reply:
It can't really be an argument out of range because we don't switch on an argument here but a class field.  Maybe `NotImplementedException` is something to consider.